packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/handles/ConnectionCreationHandles.tsx

```diff
@@ -0,0 +1,132 @@
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import { Theme, useTheme } from '@material-ui/core/styles';
+import React, { useContext, useEffect, useState } from 'react';
+import { Handle, Position } from 'reactflow';
+import { DiagramContext } from '../../contexts/DiagramContext';
+import { DiagramContextValue } from '../../contexts/DiagramContext.types';
+import { ConnectorContext } from '../connector/ConnectorContext';
+import { ConnectorContextValue } from '../connector/ConnectorContext.types';
+import { useConnector } from '../connector/useConnector';
+import { useConnectionCandidatesQuery } from './useConnectionCandidatesQuery';
+
+import { ConnectionCreationHandlesProps, ConnectionCreationHandlesState } from './ConnectionCreationHandles.types';
+
+const connectionCreationHandleStyle = (
+  position: Position,
+  theme: Theme,
+  isHovered: Position | null,
+  isMouseDown: Position | null
+): React.CSSProperties => {
+  const style: React.CSSProperties = {
+    position: 'absolute',
+    borderRadius: '0',
+    border: 'solid black',
+    borderWidth: '0 2px 2px 0',
+    display: 'inline-block',
+    padding: '2px',
+    borderColor: theme.palette.secondary.light,
+    backgroundColor: 'transparent',
+    width: 12,
+    height: 12,
+    zIndex: 9999,
+  };
+  switch (position) {
+    case Position.Left:
+      style.left = '-15px';
+      style.transform = 'translateY(-50%) rotate(135deg)';
+      break;
+    case Position.Right:
+      style.right = '-15px';
+      style.transform = 'translateY(-50%) rotate(-45deg)';
+      break;
+    case Position.Top:
+      style.top = '-15px';
+      style.transform = 'translateX(-50%) rotate(-135deg)';
+      break;
+    case Position.Bottom:
+      style.bottom = '-15px';
+      style.transform = 'translateX(-50%) rotate(45deg)';
+      break;
+  }
+  if (isHovered === position || isMouseDown === position) {
+    style.borderColor = theme.palette.primary.main;
+  }
+  return style;
+};
+
+export const ConnectionCreationHandles = ({ nodeId }: ConnectionCreationHandlesProps) => {
+  const { editingContextId, diagramId } = useContext<DiagramContextValue>(DiagramContext);
+  const { onConnectionStartElementClick } = useConnector();
+  const { setCandidates } = useContext<ConnectorContextValue>(ConnectorContext);
+
+  const [state, setState] = useState<ConnectionCreationHandlesState>({
+    shouldRender: false,
+    isHovered: null,
+    isMouseDown: null,
+  });
+
+  const candidates = useConnectionCandidatesQuery(editingContextId, diagramId, nodeId);
+  useEffect(() => {
+    const shouldRender = candidates !== null && candidates.length > 0;
+    setState((prevState) => ({ ...prevState, shouldRender }));
+    if (candidates !== null) {
+      setCandidates(candidates);
+    }
+  }, [candidates]);
```

Comment:
Just nitpicking a little here but you don't need to have `shouldRender` in the state and you don't need an useEffect to update it, it can simply be derived from the `candidates` variable. Here you still need the useEffect to update the context anyway.

```
  const candidates = useConnectionCandidatesQuery(editingContextId, diagramId, nodeId);
  const shouldRender = candidates !== null && candidates.length > 0;
  useEffect(() => {
    if (candidates !== null) {
      setCandidates(candidates);
    }
  }, [candidates]);
```

The handle would thus be rendered a tiny bit quicker since you would not need a new rendering to see the updated value of `shouldRender`. I don't really care about such minor improvement in the context of this review but in general we are way overusing `useEffect` in order to compute derived values.